src/data/data.cc

```diff
@@ -413,21 +410,72 @@ void SparsePage::Push(const SparsePage &batch) {
   }
 }
 
-void SparsePage::Push(const dmlc::RowBlock<uint32_t>& batch) {
-  auto& data_vec = data.HostVector();
+template <typename AdapterBatchT>
+uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread) {
+  // Set number of threads but keep old value so we can reset it after
+  const int nthreadmax = omp_get_max_threads();
+  if (nthread <= 0) nthread = nthreadmax;
+  int nthread_original = omp_get_max_threads();
+  omp_set_num_threads(nthread);
   auto& offset_vec = offset.HostVector();
-  data_vec.reserve(data.Size() + batch.offset[batch.size] - batch.offset[0]);
-  offset_vec.reserve(offset.Size() + batch.size);
-  CHECK(batch.index != nullptr);
-  for (size_t i = 0; i < batch.size; ++i) {
-    offset_vec.push_back(offset_vec.back() + batch.offset[i + 1] - batch.offset[i]);
+  auto& data_vec = data.HostVector();
+  size_t builder_base_row_offset = this->Size();
+  common::ParallelGroupBuilder<
+      Entry, std::remove_reference<decltype(offset_vec)>::type::value_type>
+      builder(&offset_vec, &data_vec, builder_base_row_offset);
+  // Estimate expected number of rows by using last element in batch
+  // This is not required to be exact but prevents unnecessary resizing
+  size_t expected_rows = 0;
+  if (batch.Size() > 0) {
+    auto last_line = batch.GetLine(batch.Size() - 1);
```

Comment:
Maybe sample a few more lines?

Reply:
This is not too important, it just means if we guess wrong the vector gets resized, performance is still good.